Fix jitclass method inlining and avoid unnecessary increfs

alendit opened the pull request:
This change addresses issues #2166 and #3305 . Numba wasn't able to inline jitclass' methods, because the their definitions weren't available at the compile time. When a function is lowered, the related libraries should be added using `CodeLibrary.add_linking_library`, which wasn't happening for `jitclass`. This patch adds this functionality for methods, properties and constructors.

There is one more subtle problem. Unnecessary refcounts usages are removed on the level of basic blocks. This makes sense, since analyzing across bb borders would be much more involved. However, this approach can only work after CFG is consolidated enough for inc/ref counts to be reliably grouped. At the same time, after removal of the NRT refcount calls there are additional optimizations possible, i.e. more aggressive loop unrolling. So, to achieve the best performance we would have to run the optimizations after NRT refcount removal again, with trade-off of the slower compilation. A more fine-grained approach is thinkable, with more carefully selected passes to run before and after NRT optimization, but I wasn't able to determine the perfect combination. 

Note that improving inlining also improves unnecessary NRT refcount detection since the calls are more likely to the placed in the same block.

```
@numba.jitclass([
    ('data', numba.uint32[:]),
    ('pos', numba.uint64)
])
class Container:
    def __init__(self, size = 10 ** 6):
        self.data = np.zeros(size, dtype=np.uint32)
        self.pos = 0
    
    def append(self, el):
        self.data[self.pos] = el
        self.pos += 1

@numba.njit
def append_batch():
    N = 10 ** 6
    c = Container(N)
    acc = 0
    for i in range(N):
        c.append(i)
    return c
```

For this example having an additional optimization pass makes a difference of around 30% (3.3ms vs 2.6ms). On the other hand it slows down compilation considerably. I would be glad to hear, what you would consider the best approach here.

Finally, performance numbers from the benchmark in #2166:

With current master:

```
<class 'int'>
 function_function  0.000001 0.000001 0.000001 0.000001
 method_function    0.000002 0.000002 0.000004 0.000003
 method_method      0.000164 0.001591 0.016329 0.159769
 function_method    0.000156 0.001543 0.015927 0.159938
<class 'list'>
 function_function  0.000007 0.000007 0.000007 0.000007
 method_function    0.000009 0.000008 0.000008 0.000008
 method_method      0.000174 0.001619 0.016244 0.160375
 function_method    0.000198 0.001557 0.015969 0.158437
```

With this patch

```

<class 'int'>
 function_function  0.000001 0.000001 0.000001 0.000001
 method_function    0.000001 0.000001 0.000001 0.000001
 method_method      0.000001 0.000001 0.000001 0.000001
 function_method    0.000001 0.000001 0.000001 0.000001
<class 'list'>
 function_function  0.000008 0.000007 0.000006 0.000006
 method_function    0.000008 0.000007 0.000006 0.000006
 method_method      0.000008 0.000006 0.000007 0.000006
 function_method    0.000008 0.000007 0.000007 0.000006
```

Note: this PR doesn't add any tests or benchmarks. If you have any suggestions, which part I should test and/or benchmark, I'll gladly  add it.

sklam commented on numba/jitclass/base.py:
```diff
@@ -146,6 +146,14 @@ def _fix_up_private_attr(clsname, spec):
+    """
+    Add the required libs for the callable to allow inlining.
+    """
+    cg = context.codegen()
```
Instead of adding the libraries to codegen, it should add to the currently active code library `context.active_code_library`.  See https://github.com/numba/numba/pull/3228 for details.

sklam commented on numba/jitclass/base.py:
```diff
@@ -235,6 +243,7 @@ def generic(self, args, kws):
         # Redirect resolution to __init__
         instance_type = self.key.instance_type
         ctor = instance_type.jitmethods['__init__']
+        # _add_linking_libs(self.context, ctor)
```
Dead code comment?

sklam commented on numba/targets/codegen.py:
```diff
@@ -120,8 +120,10 @@ def _optimize_final_module(self):
+
         self._codegen._mpm.run(self._final_module)
         self._final_module = remove_redundant_nrt_refct(self._final_module)
+        self._codegen._mpm.run(self._final_module)
```
I am uncomfortable with this extra optimization pass run without more benchmark data.  There have been complaints that compile time is slow for some folks.  The extra opt-pass would be too expensive for functions that are already taking minutes to compile.

sklam commented:
FYI, here's showing the difference of running optimization twice vs once on the `append_batch` code in the PR description:

Run twice:
[cfg_opt_twice.pdf](https://github.com/numba/numba/files/2634453/cfg_opt_twice.pdf)

Run once:
[cfg_opt_once.pdf](https://github.com/numba/numba/files/2634454/cfg_opt_once.pdf)

sklam commented:
As for benchmarks, can you add them to https://github.com/numba/numba-benchmark/tree/master/benchmarks?  We use ASV for benchmarking.

alendit commented:
Hi @sklam,

thanks for taking a look. I incorporated your comments and removed the second optimization pass.

Yes, the loop becomes longer because of unrolling, which increases the performance a bit, at least on my machine. It can only be done correctly after everything's inlined, so for optimal code you'd do (normal optimizations)->(remove refcnt)->(unroll loops). I absolutely agree that running whole optimization second time is not the way to go.

I'd say we need a better way to optimize inc/dec-ref call across the basic blocks. This can be seen as unrelated to this PR, although the inlining would benefit from having fewer calls inside methods.

I'll look into adding benchmarks to `numba-benchmark` (wasn't aware of that repository).

alendit commented:
@sklam I sent a PR to the benchmarking repo: https://github.com/numba/numba-benchmark/pull/8 . Just out of curiosity: do you publish the ASV results/history for numba somewhere?

sklam commented:
@alendit, they are published to http://numba.pydata.org/numba-benchmark/
